src/commenter.js

```diff
@@ -0,0 +1,164 @@
+      link: data.html_url,
+      user: data.user.login,
+    };
+  }
+
+  /**
+   * Update a comment.
+   *
+   * @param {Object} comment The object containing
+   * @param {string} content The GitHub flavored markdown to put in the comment.
+   * @return {Promise<Object>} An object containing details about the created comment.
+   */
+  async updateComment(comment, content) {
+    const url = new URL(comment.url),
+      path = url.pathname.split('/');
+    const [, , owner, repo] = path;
+    const { data } = await this.octokit.issues.updateComment({
+      // In an ideal world, updateComment would accept comment or comment's url or other field and
+      // parse these fields from there.
+      owner,
+      repo,
+      comment_id: path[path.length - 1],
+      body: content,
+    });
+    return {
+      action: 'update',
+      link: data.html_url,
+      user: data.user.login,
+    };
+  }
+
+  /**
+   * Find the comment on the given pull request with the given token.
+   *
+   * @param {string} repository The repository on which to create the comment, in slug form.
+   * @param {number|string} pullRequest The pull request number on GitHub which holds the thread
+   *   where the comment should be created.
+   * @param {string} commentToken The comment token to locate.
+   * @return {Promise<Object>} The comment object from GitHub's API.
+   */
+  async findComment({ repository, pullRequest, commentToken }) {
+    const [owner, repo] = repository.split('/');
+    const options = this.octokit.issues.listComments.endpoint.merge({
+      owner,
+      repo,
+      issue_number: pullRequest,
+    });
+    for await (const page of this.octokit.paginate.iterator(options)) {
+      for (const comment of page.data) {
+        const match = rDedupeComment.exec(comment.body);
+        if (match && match[1] === commentToken) {
+          return comment;
+        }
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Post a comment to the given pull request, and deduplicate (best-effort) comments by purpose.
+   *
+   * @param {string} repository The repository on which to create the comment, in slug form.
+   * @param {number|string} pullRequest The pull request number on GitHub which holds the thread
+   *   where the comment should be created.
+   * @param {string} content The GitHub flavored markdown to put in the comment.
+   * @param {string} purpose The purpose of the comment.
+   * @return {Promise<Object>} Some simple metadata about the action performed, suitable for
+   *   displaying a simple message detailing the action taken.
+   */
+  async postComment({ repository, pullRequest, content, purpose = null }) {
+    const commentToken = purpose && getUniqueToken({ repository, pullRequest, purpose });
+
+    const existingComment =
+      commentToken &&
+      (await this.findComment({
+        repository,
+        pullRequest,
+        commentToken,
+      }));
+
+    const taggedContent = `${getDedupeComment(commentToken)}\n${content}`;
```

Comment:
Is it ok to call `getDedupeComment` with a null `commentToken`?

Reply:
Technically not but since the comment is a comment it'll just get washed out. d6bd910